### PR TITLE
feature/add-avro-generator

### PR DESCRIPTION
**Pull request by sgadouar**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).

- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

New avro generator plugin
https://avro.apache.org/docs/current/#schemas
input: open-api specification
output: avro schemas files

### CODE REVIEW

**Comment by wing328** [comment]
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AvroCodegen.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2018 OpenAPI-Generator Contributors (https://openapi-generator.tech)
+ * Copyright 2018 SmartBear Software
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AvroCodegen.java]: @sgadouar please remove line 3 as per https://opensource.stackexchange.com/questions/7300/copyright-notice-in-the-file-header-apache-v2-license/7301#7301

File: bin/openapi3/avro-petstore.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+executable="./modules/openapi-generator-cli/target/openapi-generator-cli.jar"
+
+if [ ! -f "$executable" ]
+then
+  mvn clean package
+fi
+
+# if you've executed sbt assembly previously it will use that instead.
+export JAVA_OPTS="${JAVA_OPTS} -Xmx1024M -DloggerPath=conf/log4j.properties"
+ags="generate -i modules/openapi-generator/src/test/resources/3_0/petstore.yaml -g avro  -o samples/avro/petstore $@"
```
Comment (ackintosh) [inline bin/openapi3/avro-petstore.sh]: > -o samples/avro/petstore

How about `-o samples/openapi3/schema/petstore/avro` as the sample files are generated from openapi3 yaml?

**Comment by cbornet** [comment]
What is the use to convert OAS to Avro ? Avro only deals with the ser/deser part and not with the API/HTTP part.

**Comment by jimschubert** [comment]
@cbornet it's common for larger microservice infrastructure to define common types between services which may be consumed by teams that done own the data. Those structures may be evented through a bus like Kafka, where it would be beneficial to have a clearly defined binary transport. It may also be helpful to generate a public API structure as we do now, but then generate the avro structures for persisting the same API data for external processing via Spark. In all of these cases, Avro would be hugely beneficial.

File: bin/openapi3/avro-petstore.sh (hunk 2 above)
Comment (sgadouar) [reply bin/openapi3/avro-petstore.sh]: yes good idea !
done

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AvroCodegen.java (hunk 1 above)
Comment (sgadouar) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AvroCodegen.java]: shure, i pushed the commit

**Comment by wing328** [comment]
I'll file another PR for minor enhancements and fixes for the CircleCI failure.

**Comment by wing328** [comment]
Filed https://github.com/OpenAPITools/openapi-generator/pull/3728

**Comment by wing328** [comment]
@sgadouar thanks for the PR, which has been included in the v4.1.1 release: https://twitter.com/oas_generator/status/1165944867391860737

**Pull request merged**